ingest/src/main/java/com/microsoft/azure/kusto/ingest/ManagedStreamingIngestClient.java

```diff
@@ -112,15 +162,51 @@ public IngestionResult ingestFromStream(StreamSourceInfo streamSourceInfo, Inges
         streamSourceInfo.validate();
         ingestionProperties.validate();
 
-        if (streamSourceInfo.isLeaveOpen()) {
-            throw new UnsupportedOperationException("LeaveOpen can't be true in ManagedStreamingIngestClient");
+        UUID sourceId = streamSourceInfo.getSourceId();
+        if (sourceId == null) {
+            sourceId = UUID.randomUUID();
         }
-        streamSourceInfo.setLeaveOpen(true);
+
+        byte[] streamingBytes;
 
         try {
-            for (int i = 0; i < MAX_RETRY_CALLS; i++) {
+            streamingBytes = IOUtils.readNBytes(streamSourceInfo.getStream(), MAX_STREAMING_SIZE_BYTES + 1);
+        } catch (IOException e) {
+            throw new IngestionClientException("Failed to read from stream.", e);
+        }
+
+        // NOTE: ByteArrayInputStream's close method is a no-op, so we don't need to close it.
+        ByteArrayInputStream byteArrayStream = new ByteArrayInputStream(streamingBytes);
+
+        if (streamingBytes.length > MAX_STREAMING_SIZE_BYTES) {
+            log.info("Stream size is greater than max streaming size (%d bytes). Falling back to queued.");
+            StreamSourceInfo managedSourceInfo = new StreamSourceInfo(new SequenceInputStream(byteArrayStream, streamSourceInfo.getStream()),
+                    streamSourceInfo.isLeaveOpen(), sourceId);
+            managedSourceInfo.setCompressionType(streamSourceInfo.getCompressionType());
+            return queuedIngestClient.ingestFromStream(managedSourceInfo, ingestionProperties);
+        }
+
+        if (!streamSourceInfo.isLeaveOpen()) {
+            // From this point we don't need the original stream anymore, we cached it
+            try {
+                streamSourceInfo.getStream().close();
+            } catch (IOException e) {
+                log.warn("Failed to close stream", e);
+            }
+        }
+
+        StreamSourceInfo managedSourceInfo = new StreamSourceInfo(byteArrayStream, true, sourceId);
+        managedSourceInfo.setCompressionType(streamSourceInfo.getCompressionType());
+
+        ExponentialRetry retry = new ExponentialRetry(exponentialRetryTemplate.getMaxAttempts(), exponentialRetryTemplate.getSleepBaseSecs(),
```

Comment:
nit: I'd personally create a copy constructor:
`public ExponentialRetry(ExponentialRetry other)`

Reply:
Question - intellij says that the copy c-tor  should copy all fields, but I think we shouldn't copy "currentAttempt". WDYT?

Reply:
Yeah, without `currentAttempt`

Reply:
Maybe mark `currentAttempt` as `transient`